File: week_04/json_client.py
import requests
import json

def get_data_from(sub_url = ""):
    url = "https://json.ndtai772.repl.co" + sub_url
    print(">>Sent get request to " + url)
    response = requests.get(url)
    return response.content


print("Full data (raw): ")
full_data = get_data_from("/full")
full_data = json.loads(full_data)
print(full_data)

print("\nGet all member of team: ")
members = get_data_from("/name")
members = json.loads(members)
print(f"There are {len(members)} members:")
for mem in members:
    print(mem)

print("\nGet members' families: ")
families = get_data_from("/family")
families = json.loads(families)
for p in families:
    print("Gia dinh cua " + p["fullName"] + ": ")
    for m in p["family"]:
        print(f"Tên: {m['fullName']}, Sđt: {m['phoneNumber']}")
    print()

File: week_04/json_client.rs
use serde_json::Value;
use curl::easy::Easy;
use std::io::stdin;

fn rust_client(endpoint: &str, func: fn(Value) -> ()) {
    let mut client = Easy::new();
    let url = format!("https://json.ndtai772.repl.co{}", endpoint);
    println!("Sending get request to : {}", url);
    client.url(&url).expect("cannot get data from url");

    let mut transfer = client.transfer();
    transfer.write_function(|data| {
        func(serde_json::from_slice(data).expect("cannot parse data to json"));
        Ok(data.len())
    }).unwrap();
    transfer.perform().unwrap();
}

fn main() {
    fn get_input() -> String {
        let mut input = String::new();
        stdin().read_line(&mut input).expect("Cannot read input from stdin!");
        input
    }
    rust_client("/", |v| {
        println!("What this client could do: ");
        let endpoints = v["api-endpoint"].as_object().expect("#1");
        let jobs = endpoints.iter().enumerate().collect::<Vec<_>>();
        jobs.iter().for_each(|(i, (_k, v))| {
            println!("{}.{}", i, v.as_str().expect("#2"));
        });

        println!("What do you want? (q to quit) ");
        loop {
            let s = get_input();
            if s.trim().eq("q") {
                println!("quit!");
                return;
            }
            match s.trim().parse::<u8>() {
                Ok(0) => {
                    let (_, (k, _v)) = jobs[0];
                    rust_client(k, |data| {
                        let data = data.as_array().expect("Not an array!");
                        data.iter().filter_map(|p| p.as_object()).for_each(|people| {
                            println!("{}", people["fullName"]);
                            let family = people["family"].as_array().expect("#6");
                            println!("Gia dinh co {} nguoi: ", family.len());
                            family.iter().for_each(|p| {
                                let p = p.as_object().expect("#4");
                                println!("Tên: {}, Địa chỉ: {}, Sđt: {}", p["fullName"], p["address"], p["phoneNumber"]);
                            })
                        });
                    })
                }
                Ok(i) => {
                    if let Some((_, (k, _v))) = jobs.get(i as usize) {
                        rust_client(k, |v| { println!("{}", v); });
                    }
                }
                _ => {}
            }
        }
    });
}
